src/main/java/org/dependencytrack/event/kafka/KafkaStreamsTopologyFactory.java

```diff
@@ -78,17 +90,45 @@ Topology createTopology() {
                 }, Named.as("record_processed_vuln_scan_result"))
                 .filter((scanToken, vulnScan) -> vulnScan != null,
                         Named.as("filter_completed_vuln_scans"));
+
         completedVulnScanStream.filter((scantoken, vulnscan) -> vulnscan.getTargetType() == VulnerabilityScan.TargetType.PROJECT,
                         Named.as("filter_vuln_scans_with_project_target"))
                 .map((scantoken, vulnscan) -> {
-                    var notification = NotificationModelConverter.convert(NotificationUtil.createProjectVulnerabilityAnalysisCompleteNotification(vulnscan));
-                    return KeyValue.pair(vulnscan.getTargetIdentifier().toString(), notification);
-                }).to(KafkaTopics.NOTIFICATION_PROJECT_VULN_ANALYSIS_COMPLETE.name(),
+                    // check the failure rate and update workflow status accordingly.
+                    final double failureRate = (double) vulnscan.getScanFailed() / vulnscan.getScanTotal();
+                    try (var qm = new QueryManager()) {
+                        if (failureRate > vulnscan.getFailureThreshold()) {
+                            var vulnAnalysisState = qm.getWorkflowStateByTokenAndStep(UUID.fromString(scantoken), WorkflowStep.VULN_ANALYSIS);
+                            vulnAnalysisState.setStatus(WorkflowStatus.FAILED);
+                            vulnAnalysisState.setUpdatedAt(Date.from(Instant.now()));
+                            vulnAnalysisState.setFailureReason("Failure threshold of " + vulnscan.getFailureThreshold() + "% exceeded: " + failureRate + "% of scans failed");
+                            WorkflowState updatedState = qm.updateWorkflowState(vulnAnalysisState);
+                            qm.updateAllDescendantStatesOfParent(updatedState, WorkflowStatus.CANCELLED, Date.from(Instant.now()));
+                            vulnscan.setStatus(VulnerabilityScan.Status.FAILED);
+                            var vulnScanUpdated = qm.persist(vulnscan);
```

Comment:
This call to `persist` is currently failing with:

```
stacktrace=javax.jdo.JDOUserException: Object of type "org.dependencytrack.model.VulnerabilityScan" is detached. Detached objects cannot be used with this operation.
dt-apiserver  |         at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:698)
dt-apiserver  |         at org.datanucleus.api.jdo.JDOPersistenceManager.jdoRefresh(JDOPersistenceManager.java:495)
dt-apiserver  |         at org.datanucleus.api.jdo.JDOPersistenceManager.refresh(JDOPersistenceManager.java:507)
dt-apiserver  |         at alpine.persistence.AbstractAlpineQueryManager.persist(AbstractAlpineQueryManager.java:430)
dt-apiserver  |         at org.dependencytrack.event.kafka.KafkaStreamsTopologyFactory.lambda$createTopology$5(KafkaStreamsTopologyFactory.java:108)
```

`KafkaStreamsTopologyTest#vulnScanFailureTest` is still succeeding though, because the change has been committed [before the problematic `refresh` is called](https://github.com/stevespringett/Alpine/blob/cd6aa7ed05376935ab32bc43819eba0e3a525b7f/alpine-infra/src/main/java/alpine/persistence/AbstractAlpineQueryManager.java#L425-L432).

However, due to the failure in `refresh`, the topology will not progress past the `map_vuln_scans_with_failure_rate` processor. We can catch this scenario by extending the test, so it checks for notifications sent as well.

The `vulnScan` object has transitioned to the detached state (see [JDO Object Lifecycle](https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#lifecycle)) in `record_processed_vuln_scan_result`. It seems that the logic in Alpine's `persist` method doesn't handle *detached* objects well.

From what I can tell, calling `makePersistent` on a *detached*  object does not modify the object itself, but returns an attached object instead (see https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#pm_attach, *Copy on Attach*). `persist` does not assign the returned object to anything, so `object` stays detached. Calling `refresh` on the detached object then fails.

I'd recommend to add a new method to `QueryManager` to handle this update in an optimistic transaction (as it's done in [`recordVulnerabilityScanResult`](https://github.com/DependencyTrack/hyades-apiserver/blob/42d32b3cbabb49abd68c458ea45ef5184ce184a3/src/main/java/org/dependencytrack/persistence/QueryManager.java#L1505-L1530) already), instead of using `persist`.